### ISSUE TITLE
Values lost on commit when using `preProcessEditCellProps` with new editing API

### ISSUE THREAD

**Issue by bimbiltu** [mui/mui-x#5218]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

When using `preProcessEditCellProps` there is some strange behavior when pressing enter (or any other action that stops editing).

If you are using row edit mode:
 - Editing will stop but the text entered will be lost

If you are using cell edit mode:
 - Editing will not stop. You'll need to press enter again to commit the value and switch the cell back into its view mode.

This seems related to the `debounceMs` prop on `GridEditInputCell`. It appears that if you try to finish editing before the debounce is up, you'll run into the above behavior. If you're having trouble reproducing, increasing `debounceMs` makes it easier.

### Expected behavior 🤔

In row mode, the values should not be lost. In cell mode, pressing enter should commit the value

### Steps to reproduce 🕹

Steps:

1. Open https://codesandbox.io/s/magical-thunder-k9ynbg?file=/src/App.js
2. Begin editing the first `First Name` cell
3. Type any character and hit enter immediately after
4. Observe the above behavior

### Context 🔦

I was trying to use `preProcessEditCellProps` to validate cells and noticed this behavior.

### Your environment 🌎

See the linked codepen for package versions (they are all the latest at time of writing). I tested this in chrome and firefox and observed the same behavior in both.

For completeness, the output of the command is:
<details>
  <code>
  System:
    OS: macOS 12.2.1
  Binaries:
    Node: 12.22.12 - ~/.nvm/versions/node/v12.22.12/bin/node
    Yarn: 1.22.18 - ~/.nvm/versions/node/v12.22.12/bin/yarn
    npm: 6.14.16 - ~/.nvm/versions/node/v12.22.12/bin/npm
  Browsers:
    Chrome: 102.0.5005.115
    Edge: Not Found
    Firefox: 100.0.2
    Safari: 15.3
  npmPackages:
    @emotion/react: 11.9.3 => 11.9.0
    @emotion/styled: 11.9.3 => 11.8.1
    @mui/base:  5.0.0-alpha.84
    @mui/material: 5.8.4 => 5.8.3
    @mui/private-theming:  5.8.0
    @mui/styled-engine:  5.8.0
    @mui/system: 5.8.4 => 5.8.3
    @mui/types:  7.1.3
    @mui/utils:  5.8.0
    @mui/x-data-grid: 5.12.1 => 5.12.1
    @types/react: 16.9.51 => 16.9.51
    react: 17.0.2 => 17.0.2
    react-dom: 17.0.2 => 17.0.2
    typescript: 4.1.0-dev.20201013 => 4.1.0-dev.20201013
</code>
</details>

**Comment by m4theushw** [comment]
First, the behavior in row editing is the correct one. The rationale is that <kbd>Enter</kbd> should discard the changes if the user didn't wait for `preProcessEditCellProps` to finish. It's how the legacy API behaves without enabling the `preventCommitWhileValidating` flag: https://codesandbox.io/s/validateservernamegrid-demo-mui-x-forked-segzrs To provide feedback for the user that something is going on, we show a progress indicator when the validation takes some time. (the indicator is not shown in the example because `preProcessEditCellProps` is too fast)

Like I said, the cell editing should have the same behavior from the row editing. To achieve that, the following diff should be sufficient. We missed a call to `apiRef.current.unstable_runPendingEditCellValueMutation`, the row editing already has it:

https://github.com/mui/mui-x/blob/f347b82a16b292050e7ccaf6bf559e6b752859ae/packages/grid/x-data-grid/src/hooks/features/editRows/useGridRowEditing.new.ts#L247-L251

```diff
diff --git a/packages/grid/x-data-grid/src/hooks/features/editRows/useGridCellEditing.new.ts b/packages/grid/x-data-grid/src/hooks/features/editRows/useGridCellEditing.new.ts
index 9c7eb6ced..0ce3fa8e9 100644
--- a/packages/grid/x-data-grid/src/hooks/features/editRows/useGridCellEditing.new.ts
+++ b/packages/grid/x-data-grid/src/hooks/features/editRows/useGridCellEditing.new.ts
@@ -187,6 +187,8 @@ export const useGridCellEditing = (
     (params) => {
       const { id, field, reason } = params;
 
+      apiRef.current.unstable_runPendingEditCellValueMutation(id, field);
+
       let cellToFocusAfter: GridStopCellEditModeParams['cellToFocusAfter'];
       if (reason === GridCellEditStopReasons.enterKeyDown) {
         cellToFocusAfter = 'below';
@@ -452,7 +454,7 @@ export const useGridCellEditing = (
       updateOrDeleteFieldState(id, field, newProps);
 
       editingState = gridEditRowsStateSelector(apiRef.current.state);
-      return !editingState[id][field].error;
+      return !editingState[id]?.[field]?.error;
     },
     [apiRef, throwIfNotEditable, throwIfNotInMode, updateOrDeleteFieldState],
   );
```

Second, we could add a `dontIgnoreModificationsIfProcessingProps` (better name welcomed) prop to not ignore the modifications if `preProcessEditCellProps` is running.

```diff
diff --git a/packages/grid/x-data-grid/src/hooks/features/editRows/useGridCellEditing.new.ts b/packages/grid/x-data-grid/src/hooks/features/editRows/useGridCellEditing.new.ts
index 9c7eb6ced..a08fab465 100644
--- a/packages/grid/x-data-grid/src/hooks/features/editRows/useGridCellEditing.new.ts
+++ b/packages/grid/x-data-grid/src/hooks/features/editRows/useGridCellEditing.new.ts
@@ -198,7 +198,7 @@ export const useGridCellEditing = (
 
       let ignoreModifications = reason === 'escapeKeyDown';
       const editingState = gridEditRowsStateSelector(apiRef.current.state);
-      if (editingState[id][field].isProcessingProps) {
+      if (editingState[id][field].isProcessingProps && props.dontIgnoreModificationsIfProcessingProps) {
         // The user wants to stop editing the cell but we can't wait for the props to be processed.
         // In this case, discard the modifications.
         ignoreModifications = true;
```

**Comment by bimbiltu** [comment]
@m4theushw `preProcessEditCellProps` is synchronous here (at least from the perspective of the codepen) and never returns an error though, so there's nothing for the user to wait for

It seems that committing before `GridEditInputCell`s debounce has finished is what causes this behavior, not `preProcessEditCellProps` taking some amount of time. There's no UI indication at all that the debounce is going on (nor should there be, since it seems to be there for performance reasons and not UX). This is much easier to see if you set `debounceMs` to a high number, like `2000` and wait a moment between when you stop typing and committing.

**Comment by m4theushw** [comment]
`preProcessEditCellProps` is treated as a promise internally.

> It seems that GridEditInputCells debounce is what defines the window where this behavior occurs, and not a promise returned by preProcessEditCellProps.

Yes, without the call to `unstable_runPendingEditCellValueMutation` we missed above, it works as follow:

1. Type "foobar" in the input
2. Press <kbd>Enter</kbd> immediately
3. It commits the value "" because the debounce has not finished yet

The purpose of `unstable_runPendingEditCellValueMutation` is to not wait for the debounce time and update the value immediately.

**Comment by bimbiltu** [comment]
Got it, thanks a bunch for the more detailed explanation and quick responses @m4theushw 

Ignoring grid internals, it feels a bit unintuitive to need a `preventCommitWhileValidating` flag to get the behavior I'm looking for in the original reproducer. Maybe a documentation update could help?

To elaborate, and sorry if I'm misunderstanding how things work here, if the fix you described above were implemented and had a grid with synchronous `preProcessEditCellProps` it sounds like both of the following would hold true:
 1. If the user commits before the debounce ends (and `preventCommitWhileValidating={false}`), what they entered will be lost
 1. If the user commits some time after the debounce ends, the values entered would be saved
    - Currently, it appears the `preProcessEditCellProps` function runs whenever the `GridEditInputCell`s debounce ends naturally.

If thats the case, I think it would be very easy for developers to not run into situation 1 and assume that the grid always behaves like as perceived in situation 2. It at least took me a while to figure out there was something up with the timing of when I pressed enter.
